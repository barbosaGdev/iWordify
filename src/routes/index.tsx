import { MainNavigator } from './MainNavigator'
import { DrawerNavigator } from './DrawerNavigator'

export { MainNavigator, DrawerNavigator }
