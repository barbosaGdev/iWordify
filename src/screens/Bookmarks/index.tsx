import React, { FC, useEffect } from 'react'
import { View, Text, FlatList } from 'react-native'
import { useDispatch, useSelector } from 'react-redux'
import { DefaultRootState } from '../../store'
import { fetchBookmarks } from '../../store/actions/bookmark'
import { BookmarkState } from '../../store/reducers/bookmark'
import { ListEmptyComponent } from './elements/ListEmptyComponent'
import { RenderItem } from './elements/RenderItem'
import styles from './styles'

export const Bookmarks: FC = () => {
	const dispatch = useDispatch()

	useEffect(() => {
		dispatch(fetchBookmarks())
	}, [])

	const bookmarks = useSelector<DefaultRootState, string[]>(
		(state: { bookmark: BookmarkState }) => state.bookmark.bookmarks
	)

	return (
		<View style={styles.container}>
			<Text>Bookmarks</Text>

			<FlatList
				style={{ width: '100%' }}
				data={bookmarks}
				ListEmptyComponent={ListEmptyComponent}
				renderItem={RenderItem}
				ItemSeparatorComponent={() => <View style={styles.itemSeparator} />}
			/>
		</View>
	)
}
