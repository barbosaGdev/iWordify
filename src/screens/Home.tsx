import React from 'react'
import { View, StatusBar, Text, StyleSheet } from 'react-native'

export const Home = () => {
	return (
		<View style={styles.container}>
			<Text>Home page</Text>
		</View>
	)
}

const styles = StyleSheet.create({
	container: {
		flex: 1,
		justifyContent: 'center',
		alignItems: 'center'
	}
})
