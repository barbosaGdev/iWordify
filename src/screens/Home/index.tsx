import React, { FC, useEffect, useRef, useState } from 'react'
import { useNavigation } from '@react-navigation/native'
import { View, FlatList, LayoutRectangle, ViewToken } from 'react-native'
import { Menu } from '../../components/Menu'
import viewabilityConfig from '../../utils/viewabilityConfig'
import RenderItem, { ItemProps } from './elements/RenderItem'
import styles from './styles'
import { useDispatch } from 'react-redux'
import { fetchWordsByVowel } from '../../store/actions/dictionary'
import { DictionaryState } from '../../store/reducers/dictionary'

export type VowelsEntity = {
	vowel: string
	vowelsState: keyof DictionaryState
}

export const vowels: VowelsEntity[] = [
	{ vowel: 'A', vowelsState: 'wordsStartingWithA' },
	{ vowel: 'E', vowelsState: 'wordsStartingWithE' },
	{ vowel: 'I', vowelsState: 'wordsStartingWithI' },
	{ vowel: 'O', vowelsState: 'wordsStartingWithO' },
	{ vowel: 'U', vowelsState: 'wordsStartingWithU' }
]

export const Home: FC = () => {
	const [indexPage, setIndexPage] = useState<number | null>(0)
	const [dimensions, setDimensions] = useState<Partial<LayoutRectangle>>({})

	const dispatch = useDispatch()

	useEffect(() => {
		vowels.forEach(({ vowel, vowelsState }) => {
			dispatch(fetchWordsByVowel(vowel, vowelsState))
		})
	}, [])

	const { navigate } = useNavigation()

	const seeAboutWord = (word: string): void => {
		//@ts-ignore
		navigate('Word', { word })
	}

	const flatListRef = useRef<FlatList>(null)

	const getItemLayout = (data: any, index: number) => ({
		length: dimensions?.width! || 0,
		offset: dimensions?.width! * index || 0,
		index
	})

	const onViewableItemsChanged = async ({
		changed
	}: {
		changed: Array<ViewToken>
	}) => {
		const [actualPostBeingShown] = changed
		const index = actualPostBeingShown.index

		setIndexPage(index)
	}

	const viewabilityConfigCallbackPairs = useRef([
		{ onViewableItemsChanged, viewabilityConfig }
	])

	const renderItem: FC<{ item: VowelsEntity; index: number }> = (props) => (
		<RenderItem
			{...props}
			dimensions={dimensions}
			seeAboutWord={seeAboutWord}
		/>
	)

	return (
		<View
			onLayout={({ nativeEvent: { layout } }) => {
				setDimensions(layout)
			}}
			style={styles.container}
		>
			<Menu
				vowels={vowels}
				flatListRef={flatListRef}
				indexPage={indexPage}
				setIndexPage={setIndexPage}
			/>

			<FlatList
				data={vowels}
				ref={flatListRef}
				viewabilityConfigCallbackPairs={viewabilityConfigCallbackPairs.current}
				pagingEnabled
				renderItem={renderItem}
				showsHorizontalScrollIndicator={false}
				showsVerticalScrollIndicator={false}
				keyExtractor={(item) => item.vowel}
				getItemLayout={getItemLayout}
				horizontal
			/>
		</View>
	)
}
