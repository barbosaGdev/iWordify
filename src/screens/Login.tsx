import { useNavigation } from '@react-navigation/native'
import React, { FC, useState } from 'react'
import {
	View,
	Text,
	StyleSheet,
	TextInput,
	TouchableOpacity
} from 'react-native'
import { useDispatch } from 'react-redux'
import { authentication, AuthInput } from '../api/auth.service'
import { signIn } from '../store/actions/auth'

export const Login: FC = () => {
	const [authInput, setAuthInput] = useState<AuthInput>({
		credential: '',
		password: ''
	})

	const dispatch = useDispatch()

	const handleChange = (field: string, text: string) => {
		setAuthInput({ ...authInput, [field]: text })
	}

	const handleSubmit = () => {
		const userToken = authentication(authInput)

		userToken ? dispatch(signIn(userToken)) : console.log('error')
	}

	return (
		<View style={styles.container}>
			<View style={styles.contentContainer}>
				<Text style={styles.title}>Log In iWordify</Text>
				<TextInput
					onChangeText={(text) => handleChange('credential', text)}
					style={styles.inputs}
					placeholder='E-mail or username'
				/>
				<TextInput
					onChangeText={(text) => handleChange('password', text)}
					style={styles.inputs}
					placeholder='Password'
				/>
				<TouchableOpacity style={styles.button} onPress={handleSubmit}>
					<Text style={{ color: 'white', textAlign: 'center' }}>Access</Text>
				</TouchableOpacity>
			</View>
			<View>
				<Text>Use fingerprint</Text>
			</View>
		</View>
	)
}

const styles = StyleSheet.create({
	container: {
		flex: 1,
		alignItems: 'center',
		justifyContent: 'space-around'
	},
	contentContainer: {
		alignItems: 'center'
	},
	button: {
		marginTop: 12,
		backgroundColor: 'darkblue',
		width: 300,
		padding: 8
	},
	title: {
		fontSize: 24,
		color: '#000',
		fontWeight: 'bold'
	},
	inputs: {
		borderWidth: 1,
		borderColor: 'gray',
		width: 300,
		marginTop: 12
	}
})
