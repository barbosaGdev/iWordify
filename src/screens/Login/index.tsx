import React, { FC, useState } from 'react'
import { View, Text, TouchableOpacity } from 'react-native'
import { useDispatch } from 'react-redux'
import { AuthErrors, AuthInput, authentication } from '../../api/auth'
import { CredentialInput } from '../../components/CredentialInput'
import { PasswordInput } from '../../components/PasswordInput'
import { signIn } from '../../store/actions/auth'
import styles from './styles'

export type InputProps = {
	handleChange: (field: string, text: string) => void
}

export const Login: FC = () => {
	const [errors, setErrors] = useState<AuthErrors>()

	const [authInput, setAuthInput] = useState<AuthInput>({
		credential: '',
		password: ''
	})

	const dispatch = useDispatch()

	const handleChange = (field: string, text: string) => {
		setAuthInput({ ...authInput, [field]: text })
		setErrors({})
	}

	const validateForm = () => {
		if (!authInput.credential) {
			setErrors({ field: 'credential', message: 'Required field' })
		} else if (!authInput.password) {
			setErrors({ field: 'password', message: 'Required field' })
		} else {
			handleSubmit()
		}
	}

	const handleSubmit = () => {
		const user = authentication(authInput)

		if (user) {
			dispatch(signIn(user.username, user.token))
		} else {
			setErrors({
				status: 401,
				message: 'Unauthorized'
			})
		}
	}

	return (
		<View style={styles.container}>
			<View style={styles.contentContainer}>
				<Text style={styles.title}>Log In iWordify</Text>

				<CredentialInput handleChange={handleChange} />

				{errors?.field === 'credential' && (
					<Text style={styles.helperText}>{errors?.message}</Text>
				)}

				<PasswordInput handleChange={handleChange} />

				{errors?.field === 'password' && (
					<Text style={styles.helperText}>{errors?.message}</Text>
				)}

				<TouchableOpacity style={styles.button} onPress={validateForm}>
					<Text style={styles.buttonText}>Access</Text>
				</TouchableOpacity>
			</View>

			{errors?.status === 401 && (
				<Text style={styles.errorText}>{errors.message}</Text>
			)}
		</View>
	)
}
