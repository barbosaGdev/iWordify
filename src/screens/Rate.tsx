import React from 'react'
import { View, StatusBar, Text, StyleSheet } from 'react-native'

export const Rate = () => {
	return (
		<View style={styles.container}>
			<Text>Rate page</Text>
		</View>
	)
}

const styles = StyleSheet.create({
	container: {
		flex: 1,
		justifyContent: 'center',
		alignItems: 'center'
	}
})
