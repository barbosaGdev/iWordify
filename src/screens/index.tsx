import { Home } from './Home'
import { Rate } from './Rate'
import { Search } from './Search'
import { Word } from './Word'

export { Home, Rate, Search, Word }
