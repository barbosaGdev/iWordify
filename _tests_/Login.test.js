import React from 'react'
import renderer from 'react-test-renderer'
import thunk from 'redux-thunk'
import { createStore, combineReducers, compose, applyMiddleware } from 'redux'
import { Login } from '../src/screens'
import { Provider } from 'react-redux'
import 'react-native-gesture-handler'
import {
	auth as authReducer,
	dictionary as dictionaryReducer,
	bookmark as bookmarkReducer
} from '../src/store/reducers'

const rootReducers = combineReducers({
	auth: authReducer,
	dictionary: dictionaryReducer,
	bookmark: bookmarkReducer
})

const reducers = (state, action) => {
	return rootReducers(state, action)
}

const store = createStore(reducers, compose(applyMiddleware(thunk)))

test('renders correctly', () => {
	const tree = renderer
		.create(
			<Provider store={store}>
				<Login />
			</Provider>
		)
		.toJSON()
	expect(tree).toMatchSnapshot()
})
