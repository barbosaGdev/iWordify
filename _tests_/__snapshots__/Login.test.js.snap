// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "marginVertical": 32,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#000",
          "fontSize": 24,
          "fontWeight": "bold",
          "marginVertical": 24,
        }
      }
    >
      Log In iWordify
    </Text>
    <View
      style={
        Object {
          "borderColor": "gray",
          "borderWidth": 1,
          "flexDirection": "row",
          "marginTop": 12,
          "paddingHorizontal": 12,
          "width": 350,
        }
      }
    >
      <TextInput
        onChangeText={[Function]}
        placeholder="E-mail or username"
      />
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "borderColor": "gray",
          "borderWidth": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginTop": 12,
          "paddingHorizontal": 12,
          "width": 350,
        }
      }
    >
      <TextInput
        onChangeText={[Function]}
        placeholder="Password"
        secureTextEntry={true}
      />
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <Text
          allowFontScaling={false}
          selectable={false}
          style={
            Array [
              Object {
                "color": undefined,
                "fontSize": 20,
              },
              undefined,
              Object {
                "fontFamily": "anticon",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          îš‡
        </Text>
      </View>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "darkblue",
          "marginTop": 12,
          "opacity": 1,
          "padding": 8,
          "width": 350,
        }
      }
      testID="LoginButton"
    >
      <Text
        style={
          Object {
            "color": "white",
            "textAlign": "center",
          }
        }
      >
        Access
      </Text>
    </View>
  </View>
</View>
`;
